apis/python/src/tiledbsoma/io/registration/ambient_label_mappings.py

```diff
@@ -0,0 +1,402 @@
+            if "raw" in previous.var_axes:
+                var_maps["raw"] = dict(previous.var_axes["raw"].data)
+
+        else:
+            # One input may not have a raw while the next may have one
+            raw_var_map = {}
+            raw_var_next_soma_joinid = 0
+            if "raw" in previous.var_axes:
+                raw_var_axis = previous.var_axes["raw"]
+                raw_var_map = raw_var_axis.data
+                raw_var_next_soma_joinid = raw_var_axis.get_next_start_soma_joinid()
+            raw_var_ids = get_dataframe_values(adata.raw.var, var_field_name)
+            for raw_var_id in raw_var_ids:
+                if raw_var_id not in raw_var_map:
+                    raw_var_map[raw_var_id] = raw_var_next_soma_joinid
+                    raw_var_next_soma_joinid += 1
+
+            var_maps["raw"] = raw_var_map
+
+        tiledbsoma.logging.logger.info(
+            f"Registration: accumulated to nobs={len(obs_map)} nvar={len(var_map)}."
+        )
+        return cls(
+            obs_axis=AxisAmbientLabelMapping(data=obs_map, field_name=obs_field_name),
+            var_axes={
+                ms_name: AxisAmbientLabelMapping(data=data, field_name=var_field_name)
+                for ms_name, data in var_maps.items()
+            },
+        )
+
+    @classmethod
+    def from_h5ad_append_on_experiment(
+        cls,
+        h5ad_file_name: str,
+        previous: Self,
+        *,
+        measurement_name: str,
+        obs_field_name: str = "obs_id",
+        var_field_name: str = "var_id",
+    ) -> Self:
+        """Extends registration data to one more H5AD input file."""
+        tiledbsoma.logging.logger.info(f"Registration: registering {h5ad_file_name}.")
+
+        adata = ad.read_h5ad(h5ad_file_name, "r")
+
+        return cls.from_anndata_append_on_experiment(
+            adata,
+            previous,
+            measurement_name=measurement_name,
+            obs_field_name=obs_field_name,
+            var_field_name=var_field_name,
+        )
+
+    @classmethod
+    def from_h5ad_appends_on_experiment(
+        cls,
+        experiment_uri: Optional[str],
+        h5ad_file_names: Sequence[str],
+        *,
+        measurement_name: str,
+        obs_field_name: str,
+        var_field_name: str,
+        context: Optional[SOMATileDBContext] = None,
+    ) -> Self:
+        """Extends registration data from the baseline, already-written SOMA
+        experiment to include multiple H5AD input files."""
+
+        if experiment_uri is not None and tiledbsoma.Experiment.exists(experiment_uri):
+            registration_data = cls.from_isolated_soma_experiment(
+                experiment_uri,
+                obs_field_name=obs_field_name,
+                var_field_name=var_field_name,
+                context=context,
+            )
+        else:
+            registration_data = cls(
+                obs_axis=AxisAmbientLabelMapping(data={}, field_name=obs_field_name),
```

Comment:
Optional suggestion: if passing `data={}` in is a thing you do routinely, you could make `data: Dict[...] = attrs.field(factory=dict)` so that by default you get a new empty dict as the value and you don’t need to explicitly provide it if empty.

Reply:
🔥

Reply:
@thetorpedodog thanks!! Awesome to know.

In this case it's highly localized, not a pattern -- so I think I'll leave it as-is